plutus-contract/src/Language/Plutus/Contract/Effects/WatchAddress.hs

```diff
@@ -31,32 +36,98 @@ import           Language.Plutus.Contract.Util              (loopM)
 type AddressSymbol = "address"
 
 type HasWatchAddress s =
-    ( HasType AddressSymbol (Address, Tx) (Input s)
-    , HasType AddressSymbol Address (Output s)
+    ( HasType AddressSymbol AddressChangeResponse (Input s)
+    , HasType AddressSymbol AddressChangeRequest (Output s)
     , ContractRow s)
 
-type WatchAddress = AddressSymbol .== ((Address, Tx), Address)
+type WatchAddress = AddressSymbol .== (AddressChangeResponse, AddressChangeRequest)
+
+-- | Information about transactions that spend or produce an output at
+--   an address. The transactions are returned in the order they appear
+--   in the ledger (oldest first).
+data AddressChangeResponse =
+    AddressChangeResponse
+        { acrAddress :: Address
+        , acrSlot    :: Slot
+        , acrTx      :: Tx
+        }
+        deriving stock (Eq, Generic, Show)
+        deriving anyclass (ToJSON, FromJSON)
+
+instance Pretty AddressChangeResponse where
+    pretty AddressChangeResponse{acrAddress, acrTx, acrSlot} =
+        hang 2 $ vsep
+            [ "Address:" <+> pretty acrAddress
+            , "Slot:" <+> pretty acrSlot
+            , "TxId:" <+> pretty (txId acrTx)
+            ]
+
+data AddressChangeRequest =
+    AddressChangeRequest
+        { acreqSlot    :: Slot
+        , acreqAddress :: Address
+        , acreqTxId    :: Maybe TxId
```

Comment:
Likewise, not really sure what this means.

Reply:
Ah, it's below. I'd probably put it in the haddock for this datatype and reference it in the function haddock.